### PR TITLE
Parse tags starting with digits and containing spaces (close #1072)

### LINKED ISSUE #1072: Allow tags to start with a numeral
Starting tags with a numeral kinda works, but the tag isn't highlighted in blue showing its recognised as a tag, and autocomplete doesn't work. Queries based on the tag still work though.
My use-case is that I'm following the filing system 'Johnny Decimal' which allows creating consistent filing across areas like notes, email, electronic folders, and physical folders. But this requires that the name starts with a numeric id.

### PR DESCRIPTION
**Pull request by Maarrk**
I'd suggest to continue discussion in #1072

### CODE REVIEW

File: plug-api/lib/tags.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import {
   type ParseTree,
   traverseTree,
 } from "@silverbulletmd/silverbullet/lib/tree";
+import { hashtagName } from "$common/markdown_parser/parser.ts";
```
Comment (zefhemel) [inline plug-api/lib/tags.ts:7]: The idea of `plug-api/*` is that these can be liberally imported from plugs without pulling in too much random code. In this case you're importing basically the entire markdown parser + all lezer parser code, not a good idea. Maybe you can move the `hashtagName` and related functions even to this very file (tags.ts) and import them into the markdown parser instead?

File: common/markdown_parser/parser.ts [hunk 2]
```diff
@@ -594,6 +594,32 @@ const Hashtag = regexParser(
...
+    } else { // this is correct #<name>
+      return text.slice(2, -1);
+    }
+  } else { // this is just #name
+    return text.slice(1);
+  }
+}
+
+/** Get markup for a hashtag name with # prefix and angle brackets if necessary */
+export function hashtagText(name: string): string {
```
Comment (zefhemel) [inline common/markdown_parser/parser.ts:614]: What about `renderHashtag`?

File: common/markdown_parser/parser.ts [hunk 3]
```diff
@@ -594,6 +594,32 @@ const Hashtag = regexParser(
   },
 );
 
+/** Extract the name from hashtag text, removing # prefix and <angle brackets> if necessary */
+export function hashtagName(text: string): string {
```
Comment (zefhemel) [inline common/markdown_parser/parser.ts:598]: what about `extractHashtag`?

File: common/markdown_parser/parser.ts [hunk 4]
```diff
@@ -594,6 +594,32 @@ const Hashtag = regexParser(
...
+    }
+  } else { // this is just #name
+    return text.slice(1);
+  }
+}
+
+/** Get markup for a hashtag name with # prefix and angle brackets if necessary */
+export function hashtagText(name: string): string {
+  // detect with the same regex as the parser
+  const simple: string = "#".concat(name);
```
Comment (zefhemel) [inline common/markdown_parser/parser.ts:616]: What about simply `const simple = "#" + name"`?

**Review by zefhemel — CHANGES_REQUESTED** [review]
Added a few comments, but looking forward to get this in!

File: common/markdown_parser/parser.ts (hunk 4 above)
Comment (Maarrk) [reply common/markdown_parser/parser.ts]: Changed. I had had issues with `+` for strings in typescript, still don't know where I can and can't use it.

File: plug-api/lib/tags.ts (hunk 1 above)
Comment (Maarrk) [reply plug-api/lib/tags.ts]: That's an area in which I have very little experience, and I'm glad to learn. Now everything is importing `markdown_parser/constants.ts` instead. Suddenly it makes sense why this module exists :smile:

File: plug-api/lib/tags.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@ import {
   type ParseTree,
   traverseTree,
 } from "@silverbulletmd/silverbullet/lib/tree";
+import { tagRegex } from "$common/markdown_parser/constants.ts";
```
Comment (zefhemel) [inline plug-api/lib/tags.ts:7]: The problem we have now is that this means that `$common` needs to be defined in any plug relying on this file, which will be the case for all the builtins, but not for external plugs. Two options: either move the constant into plug-api/lib (again: I think this very file tags.ts would be fine) or use a relative path: `../../common/markdown_parser/...` instead.

**Comment by zefhemel** [comment]
Ok you know what, let me just merge this as it and fix it on main.

**Comment by Maarrk** [comment]
OK, thanks! I'll write down the changes I needed to do here for future PRs

**Comment by zefhemel** [comment]
I hope you're taking a note in SilverBullet.

**Pull request merged**